### ISSUE TITLE
Add docs for publishing pacts to broker

### ISSUE THREAD

**Issue by bethesque** [pact-foundation/pact-python#137]
I was just adding the pact-python API token configuration docs to https://docs.pactflow.io/docs/getting-started/#configuring-your-api-token and I realised that pact publishing for python is done by manual PUTing.

There is a pact publication task in the underlying pact-ruby-standalone. Can we wire it up to the python code? Ideally, it would just be a "blindly pass all the arguments through" kind of deal. I don't think there's any need to wrap an extra python CLI around it.

The documentation for the CLI is here: https://github.com/pact-foundation/pact_broker-client#usage---cli

**Comment by elliottmurray** [comment]
Where are you seeing that?

https://github.com/pact-foundation/pact-python/blob/48f0f513686228fdde15a12ebac32f821efccd36/pact/pact.py#L179

**Comment by elliottmurray** [comment]
@bethesque forgot to tag you to the above.

**Comment by bethesque** [comment]
Ah! That's good to know. It's just missing the documentation then. There's no word 'publish' on the README.

Here was the manual upload. It's probably just really old code. https://github.com/pact-foundation/pact-python/blob/48f0f513686228fdde15a12ebac32f821efccd36/examples/e2e/tests/test_user_consumer.py#L66

**Comment by elliottmurray** [comment]
Yes I think I need to update some of the docs with a working example of this.

**Comment by bethesque** [comment]
We're getting people using pact-python who are doing manual PUT requests to the broker because they don't know that pact python can do it for them.

**Comment by alecgerona** [comment]
In addition to this, can we get some clarifications on how to publish verifications that's _not_ on the cli? The line that apparently does this is commented out for some reason. 

https://github.com/pact-foundation/pact-python/blob/40de3e34ca88336bd5a5b791e6eba9aae9feffae/pact/verify_wrapper.py#L168

As a workaround, I think I'd have to construct my own command, instantiate `VerifyWrapper()` and call `publish_results()` myself. This is a huge blocker for us trying to integrate Pact for our microservices built in Python using `pytest` as our runner.

**Comment by elliottmurray** [comment]
That should be deleted. It was a hang up from a refactor I did recently. I wasn't aware there was actually a publish example. Check out in examples:

https://github.com/pact-foundation/pact-python/blob/40de3e34ca88336bd5a5b791e6eba9aae9feffae/examples/e2e/tests/consumer/test_user_consumer.py#L52

**Comment by alecgerona** [comment]
I'm referring to publishing the verification results in the provider test. Is this really ideally done via cli alone?

**Comment by elliottmurray** [comment]
Ah sorry. So that refactor was a first attempt at providing a native Python experience rather than just the CLI. I haven't finished that (I have been stumped on a CI issue for the examples) It might 'just work'. You can look at the unit test in test_verifier.py and if you look at the docs/verifier_docs_examples branch you can see me putting the example together (though it doesn't yet have the publish option) This is still wip and I've been distracted by some other tasks.

**Comment by alecgerona** [comment]
All good. I think I'll go with the pact-provider-verifier cli route for now. I'd love to open a PR to augment the docs. I think it's still too high-level in some regards.

**Comment by bethesque** [comment]
What's the status of this peeps? Is there an interface for publishing pacts? Got some people having issues because they're trying to call the API directly and it's confusing.

**Comment by mikahjc** [comment]
I wrote the code here that hooks into the Ruby CLI, I can probably update the README to reference how to use it as well. But it works just as you suggested @bethesque, grab all the arguments and pass them through. It's been a while, but I believe it was set up with the additional parameters passed in when you call `.has_pact_with()`:
https://github.com/pact-foundation/pact-python/blob/48f0f513686228fdde15a12ebac32f821efccd36/pact/pact.py#L77

**Comment by bethesque** [comment]
What's the status of this feature? Trying to work out if I need to raise a request to add the branch to the publishing, but I still can't see any docs for this in the readme. Is it supported? Or do you just tell people to publish with the pact broker client CLI?

**Comment by mikahjc** [comment]
AFAIK this still works, but I'm no longer with the company where I wrote the integration. Let me double check, and if it does work I'll put in a PR to get this added to the readme.

**Comment by bethesque** [comment]
Thanks. TBH, we'd actually prefer everyone used the Pact CLI directly, as it makes rolling out changes like the one I mentioned super simple, rather than having to update code/docs each time. Perhaps we just don't mention the Python one in the docs, and start pointing people at the standard CLI. This is going to be even more relevant when Pact Python moves off the pact-ruby-standalone onto Rust (which doesn't have the CLI bundled in it).

**Comment by mikahjc** [comment]
Sorry for the late follow-up, but just to confirm, even though the code exists in here to publish pacts, we don't want it documented because the current direction is to have people use the CLI instead?

> This is going to be even more relevant when Pact Python moves off the pact-ruby-standalone onto Rust (which doesn't have the CLI bundled in it).

If this is the case, should this feature be removed? It depends on the Ruby CLI, so once Pact Python moves off pact-ruby-standalone, this will likely break.

**Comment by bethesque** [comment]
It makes sense to remove it when the rust backed version becomes the official release I think. I'm guessing that will be a major version release. What do you think?

**Comment by mefellows** [comment]
See also [this conversation](https://github.com/pact-foundation/pact-js/issues/765) at the Pact JS repo. 

TL;DR - it's handy having the CLI bundled with the Python interface (because it will still be the recommended way to communicate to the Pact Broker, publish etc. and saves an extra installation step), but I would still recommend removing the Python wrapper.

**Comment by mikahjc** [comment]
I agree. I think at this point it's more a decision of do it when the rust backed version becomes official, or just remove it and be done with it.

**Comment by YOU54F** [comment]
Docs exist for publishing via the pact-python api

https://github.com/pact-foundation/pact-python#uploading-pact-files-to-a-pact-broker

will table this discussion about removing it from the pact-python core when we move to the pact ffi, and whether pythonistas want to maintain a publishing interface over the pact broker client (as per pact-js's intentions in https://github.com/pact-foundation/pact-js-core/pull/435 and with https://github.com/pact-foundation/pact-js-cli)

essentially separating the concerns. the ffi interface would still need to talk to a broker to get pacts, and publish verification results through already exposed methods in the ffi.

**Comment by YOU54F** [comment]
@alecgerona I want to hear more about this!!! ❤️ 

>For more context, I'm just trying to reproduce the sample CI workshop you guys have for javascript.
